core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/Histogram.java

```diff
@@ -103,13 +107,27 @@ public int compare(Histogram.Bucket b1, Histogram.Bucket b2) {
             }
         });
 
+        public static final OrderSpec SPEC = new OrderSpec() {
```

Comment:
I see. So there will only be 4 types of orders supported:
1. `_key`: order by bucket key (`_term` and `_time` would be a deprecated alias to this)
2. `_count`: order by bucket doc count
3. order by aggregation (assume this sort if order key does not match any of the above reserved sort keys)
4. compound: sort by a list of any combination of the sorts above (`order` param is an array)

If it is safe to assume this model will fit the current and future agg order features, then I can move much more the code into the common `BucketOrder` class. I may need to fight with the generics a bit (used for the `Comparator`) but I don't think its too bad.

Good idea with serialization logic checking the version. This way I can standardize the IDs for the 4 sorts above and remove that `OrderSpec` interface. Thanks for the input.